src/network/networkd-wifi-wpa_supplicant.c

```diff
@@ -0,0 +1,375 @@
+
+        for (;;)
+                if (ssid) {
+                        r = sd_bus_message_skip(m, "{sv}");
+                        if (r < 0)
+                                return bus_log_parse_error(r);
+                        if (r == 0)
+                                break;
+                } else {
+                        r = sd_bus_message_enter_container(m, 'e', "sv");
+                        if (r < 0)
+                                return bus_log_parse_error(r);
+                        if (r == 0)
+                                break;
+
+                        r = sd_bus_message_read(m, "s", &name);
+                        if (r < 0)
+                                return bus_log_parse_error(r);
+
+                        if (streq(name, "ssid")) {
+                                r = sd_bus_message_read(m, "v", "s", &ssid);
+                                if (r < 0)
+                                        return bus_log_parse_error(r);
+                        } else
+                                r = sd_bus_message_skip(m, "v");
+
+                        r = sd_bus_message_exit_container(m);
+                        if (r < 0)
+                                return bus_log_parse_error(r);
+                }
+
+        r = sd_bus_message_exit_container(m);
+        if (r < 0)
+                return bus_log_parse_error(r);
+
+        r = sd_bus_message_exit_container(m);
+        if (r < 0)
+                return bus_log_parse_error(r);
+
+        if (*ssid == '"')
+                ssid++;
+
+        n = strlen(ssid);
+        if (ssid[n-1] == '"')
+                ssid = strndupa(ssid, n-1);
+
+        if (streq_ptr(link->ssid, ssid))
+                return 0;
+
+        r = free_and_strdup(&link->ssid, ssid);
+        if (r < 0)
+                return log_oom();
+
+        log_link_info(link, "Connected to %s", ssid);
+
+        (void) link_reconfigure(link);
+        return 0;
+}
+
+static int wpa_supplicant_get_network_properties(Link *link) {
+        _cleanup_(sd_bus_slot_unrefp) sd_bus_slot *slot = NULL;
+        int r;
+
+        assert(link);
+        assert(link->wpa_supplicant_network_path);
+
+        r = sd_bus_call_method_async(
+                        link->manager->bus,
+                        &slot,
+                        "fi.w1.wpa_supplicant1",
+                        link->wpa_supplicant_network_path,
+                        "org.freedesktop.DBus.Properties",
+                        "Get",
+                        wpa_supplicant_get_network_properties_handler,
+                        link,
+                        "ss",
+                        "fi.w1.wpa_supplicant1.Network",
+                        "Properties");
+        if (r < 0)
+                return log_link_error_errno(link, r, "Failed to get wifi network properties: %m");
+
+        (void) sd_bus_slot_set_destroy_callback(slot, (sd_bus_destroy_t) link_netlink_destroy_callback);
```

Comment:
hmm, if we can't set that we sould probably have a clean-up code paths

(or, if this realistically cannot fail, it should be enclosed in assert_se())